extensions/wikia/UserRenameTool/SpecialRenameuser_body.php

```diff
@@ -32,99 +31,160 @@ public function execute( $par ) {
 		$this->setHeaders();
 		$this->checkPermissions();
 
-		$out = $this->getOutput();
-		$this->getOutput()->addScript(
-			Html::linkedScript(
-				AssetsManager::getInstance()->getOneCommonURL( '/extensions/wikia/UserRenameTool/js/NewUsernameUrlEncoder.js' )
-			)
-		);
-
 		if ( wfReadOnly() || !$wgStatsDBEnabled ) {
-			$out->readOnlyPage();
+			$this->getOutput()->readOnlyPage();
+
 			return;
 		}
 
-		// Get the request data
-		$request = $this->getRequest();
-		$oldUsername = $request->getText( 'oldusername', $par );
-		$newUsername = $request->getText( 'newusername' );
-		$reason = $request->getText( 'reason' );
-		$token = $this->getUser()->getEditToken();
-		$notifyRenamed = $request->getBool( 'notify_renamed', false );
-		$confirmAction = false;
-
-		if ( $request->wasPosted() && $request->getInt( 'confirmaction' ) ) {
-			$confirmAction = true;
+		$user = $this->getUser();
+
+		// TODO: remove after QA tests
+		if ( static::shouldUnlock() ) {
+			$user->setGlobalFlag( 'wasRenamed', false );
+			$user->saveSettings();
 		}
+		// TODO: end
 
-		$warnings = [];
-		$errors = [];
-		$info = [];
-
-		if ( $this->validRenameRequest() ) {
-			$process = new RenameUserProcess( $oldUsername, $newUsername, $confirmAction, $reason, $notifyRenamed );
-			$status = $process->run();
-			$warnings = $process->getWarnings();
-			$errors = $process->getErrors();
-			if ( $status ) {
-				$info[] = $this->msg( 'userrenametool-info-in-progress' )->inContentLanguage()->escaped();
-			}
+		if ( \RenameUserHelper::canUserChangeUsername( $user ) ) {
+			$this->renderForm( $user );
+		} else {
+			$this->renderDisallow();
 		}
 
-		$showConfirm = ( empty( $errors ) && empty( $info ) );
+		return;
+	}
 
-		// note: errors and info beyond this point are non-blocking
+	public static function validateData( array $data, User $user ) {
+		$errorList = [];
+		$newUsername = $data['newUsername'];
 
-		if ( !empty( $oldUsername ) ) {
-			$oldUser = User::newFromName( $oldUsername );
-			if ( $oldUser->getGlobalFlag( 'requested-rename', 0 ) ) {
-				$info[] = $this->msg( 'userrenametool-requested-rename', $oldUsername )->escaped();
-			} else {
-				$errors[] = $this->msg( 'userrenametool-did-not-request-rename', $oldUsername )->escaped();
-			}
-			if ( $oldUser->getGlobalFlag( 'wasRenamed', 0 ) ) {
-				$errors[] = $this->msg( 'userrenametool-previously-renamed', $oldUsername )->escaped();
+		if ( $data['token'] === '' ) {
+			$errorList[] = 'userrenametool-error-token-not-exists';
+		}
+
+		if ( $newUsername === '' ) {
+			$errorList[] = 'userrenametool-error-no-username';
+		}
+
+		if ( $newUsername !== $data['newUsernameRepeat' ] ) {
+			$errorList[] = 'userrenametool-error-not-repeated-correctly';
+		}
+
+		if ( !ctype_alnum( $newUsername ) ) {
+			$errorList[] = 'userrenametool-error-non-alphanumeric';
```

Comment:
Should we use MW user name validation here?

Reply:
Rules here are more restrictive than in User class implementation. I mean
`a username that contains non-alphabetic characters, for example, card suit symbols or trademark signs`.

So the question is: should we move this to RenameUserProcess? Because we need additional validation above the standard one. Correct me if I'm wrong.

Reply:
@leshniak I think we should make our validation logic consistent. If we are to limit user renames to names that only contain alphanumeric characters, we should update user name validation on registration accordingly to apply the same restriction to new registrations going forward.

Reply:
@TK-999 and @leshniak — please, talk to @ludwikkazmierczak. Per our conversation, both registration and rename tool should single source for name validation. This is a requirement. Make sure it is met. Otherwise, with two separate validation mechanisms, we are creating unnecessary technical debt.

Reply:
Ok, fixed now, my mistake.